### PR TITLE
Patch anatolii borysov/tolerate service failures

### PR DESCRIPTION
**Pull request by jongiddy**
## High-level description

Changes to winpanda to:
 

1. implement operation retrying helper tools
2. rework the winpanda "start" command to support retrying logic for service start operation
3. change logic of the winpanda "start" command for not to fail entire chain of service start commands, if a single service fails to start

This is slightly updated version of #6628 since changes cannot be made to it without access to the original repo.

## Corresponding DC/OS tickets (required)

  - [DCOS_OSS-5731](https://jira.mesosphere.com/browse/DCOS_OSS-5731) Winpanda App: Tolerate Service Start Failures

## Related tickets (optional)

  - [DCOS-ID](https://jira.mesosphere.com/browse/DCOS-<number>) JIRA title / short description.

## Checklist for component/package updates:

If you are changing components or packages in DC/OS (e.g. you are bumping the sha or ref of anything underneath `packages`), then in addition to the above please also include:

  - [ ] Change log from the last version integrated (this should be a link to commits for easy verification and review): [example](https://github.com/dcos/dcos-mesos-modules/compare/f6fa27d7c40f4207ba3bb2274e2cfe79b62a395a...6660b90fbbf69a15ef46d0184e36755881d6a5ae)
  - [ ] Included a test which will fail if code is reverted but test is not. If there is no test please explain.
  - [ ] Test Results: [link to CI job test results for component]
  - [ ] Code Coverage (if available): [link to code coverage report]

### CODE REVIEW

**Comment by jongiddy** [comment]
@mesosphere-mergebot label Ready For Review

**Comment by jongiddy** [comment]
@mesosphere-mergebot changelog-not-required Update to unreleased code

**Comment by jongiddy** [comment]
@mesosphere-mergebot request-review @DominikDary @tillt

File: packages/winpanda/extra/src/winpanda/common/utils.py [hunk 1]
```diff
@@ -128,3 +130,65 @@ def run_external_command(cl_elements, timeout=30):
...
+    retrying loop.
+
+    :param attempt:             int, attempt number
+    :param retry_interval_base: float, lower limit of the retry interval
+    :param retry_interval_cap:  float, upper limit of the retry interval
+
+    :return:                    float, retry interval in seconds
+    """
+    # Implementation of the 'Exponential Backoff' with 'Full Jitter'.
+    # Ref: http://www.awsarchitectureblog.com/2015/03/backoff.html
```
Comment (tillt) [inline packages/winpanda/extra/src/winpanda/common/utils.py]: Thanks for adding the link.

File: packages/winpanda/extra/src/winpanda/common/utils.py [hunk 2]
```diff
@@ -128,3 +130,65 @@ def run_external_command(cl_elements, timeout=30):
...
+    return random.uniform(
+        retry_interval_base,
+        min(retry_interval_cap, pow(2, attempt) * retry_interval_base)
+    )
+
+
+def retry_on_exc(exceptions=(Exception,), max_attempts=0,
+                 retry_interval_base=0.5, retry_interval_cap=5.0):
+    """Apply retrying logic to a function/method being decorated.
+    """
```
Comment (tillt) [inline packages/winpanda/extra/src/winpanda/common/utils.py]: That is super neat - but I am way over my head with this python magic.

**Pull request merged**